mozilla-ai/lumigator
# Ground Truth Service

## Description
This PR is part of a stacked PR that adds a Ground Truth Service to spin up (and later, down) a Ray cluster to perform ground truth generation.  

In the interest of not having too much code to review since it's already getting complex, I'd like to break this PR up into several parts: 

1. A skeleton for a ground truth service. <-- this PR
2. The ability to run the working service to generate the dataset 
3. Saving the ground truth evaluation to file storage. <-- the goal for this sprint
4. The ability to delete/update deployments
5. Unit tests! <- might come before deletes

## Jira Tickets: 

- https://mzai.atlassian.net/browse/MZPLATFORM-88 

As a user of the MzAI Platform, when I run an evaluation experiment,
I want ‘ground truth’ to be  automatically generated for my dataset when it is missing,
So that I can run an evaluation experiment without having to manually provide ground-truth examples

## This PR specifically: 

1. Creates a route `/api/v1/ground-truth/deployments` to list and create deployments

<img width="826" alt="Screenshot 2024-07-01 at 12 52 11 PM" src="https://github.com/mozilla-ai/mzai-platform/assets/3837836/339e2003-2093-4d1d-b900-28709ca7aea9">

2. This spins up a new deployment in the local Ray cluster in docker-compose at `http://127.0.0.1:8265/#/overview`. You can see the deployment doesn't finish spinning up because we don't have PAT access 


<img width="1767" alt="Screenshot 2024-07-01 at 12 54 28 PM" src="https://github.com/mozilla-ai/mzai-platform/assets/3837836/c6623c4b-740e-474a-b990-e37bd529315a">


## To test code locally: 

1.   pull branch and go to root
2. `make local-up` to start the server
3. In curl, run:  

```
curl -X POST http://127.0.0.1/api/v1/ground-truth/deployments \
  -H "Content-Type: application/json" \
  -d '{
    "deployment_id": "fac6108e-112c-46ca-ae88-92de6857be67",
    "deployment_type": "groundtruth",
    "args": {},
    "description": "My test app",
    "name": "Summarizer"
  }'
{"id":"5d7488e7-885c-4d7f-82e6-ea65ff057268","name":"summarizer","description":"My test app","status":"created","created_at":"2024-07-01T19:17:00.056574Z","updated_at":null}%
```

4. In the browser, you can access info about the ground truth service at `http://127.0.0.1/api/v1/ground-truth/deployments`. You should see: 

<img width="527" alt="Screenshot 2024-07-01 at 12 46 09 PM" src="https://github.com/mozilla-ai/mzai-platform/assets/3837836/c62ca9bd-70b1-4037-b67a-0cc3de35b94f">

5. You should see the deployment try to spin up in the Ray `cluster` in `serve` `http://127.0.0.1:8265/#/serve`

Just an FYI @imtihan @macaab26
